FloripaPublicTransportation/MapViewController.swift

```diff
@@ -0,0 +1,140 @@
+    /// during the lifetime of `MapViewController`
+    private func showAndUpdateUserLocation() {
+        self.locationManager.startUpdatingLocation()
+        self.mapView.showsUserLocation = true
+    }
+    
+    /// NSTimer callback to show the pin callout automatically once it is created
+    internal func showPinCallout() {
+        // can't be private otherwise the selector does not recognize it
+        mapView.selectAnnotation(mapView.annotations.filter({ (a: MKAnnotation) -> Bool in
+            return !a.isKindOfClass(MKUserLocation) }).first!, animated: true)
+    }
+    
+    // MARK: - Delegates
+    
+    // MARK: CLLocationManagerDelegate
+    
+    func locationManager(manager: CLLocationManager, didChangeAuthorizationStatus status: CLAuthorizationStatus) {
+        if status == CLAuthorizationStatus.AuthorizedAlways || status == CLAuthorizationStatus.AuthorizedWhenInUse {
+            self.showAndUpdateUserLocation()
+        }
+    }
+    
+    // MARK: MKMapViewDelegate
+    
+    func mapView(mapView: MKMapView, viewForAnnotation annotation: MKAnnotation) -> MKAnnotationView? {
+        guard !annotation.isKindOfClass(MKUserLocation) else {
+            return nil
+        }
+        
+        let annotationView = MKPinAnnotationView()
+        annotationView.annotation = annotation
+        annotationView.canShowCallout = true
+        annotationView.pinTintColor = UIColor.redColor()
+        
+        let okButton = UIButton(type: .System)
+        okButton.setTitle("Ok", forState: .Normal)
+        okButton.frame = CGRectMake(0, 0, annotationView.frame.width, annotationView.frame.height)
+        annotationView.rightCalloutAccessoryView = okButton
+        
+        return annotationView
+    }
+    
+    func mapView(mapView: MKMapView, didAddAnnotationViews views: [MKAnnotationView]) {
+        /// zooms the map to the user location on first load
+        if mapView.annotations.count == 1 && mapView.annotations.first!.isKindOfClass(MKUserLocation) {
+            mapView.showAnnotations(mapView.annotations, animated: true)
+            return
+        }
+        else if mapView.annotations.count >= 1  {
+            // for some reason, the callout is not shown if the line below is called here:
+            //mapView.selectAnnotation(mapView.annotations.last!, animated: false)
+            
+            // will have to adopt an ugly way and hardcode a time interval that always works:
+            _ = NSTimer.scheduledTimerWithTimeInterval(0.8, target: self, selector: "showPinCallout", userInfo: nil, repeats: false)
+        }
+    }
+    
+    func mapView(mapView: MKMapView, annotationView view: MKAnnotationView, calloutAccessoryControlTapped control: UIControl) {
+        guard self.navigationController != nil else {
+            return
+        }
+        
+        if let routesListVC = self.navigationController!.viewControllers[0] as? RoutesListViewController {
```

Comment:
It definitely would, however as I have commented inside the MKMapViewDelegate function "didAddAnnotationViews", the callout would not be opened when called from there. And I knew this was ugly, but after searching I could only find people claiming this was Apple's bugs and suggesting the ugly timer everywhere - so I felt a little less uncomfortable with that. However as I said in a recent reply here, maybe this would have been fixed if I had added the annotation inside the geocoder completion handler.